crates/datasources/src/common/sink/lance.rs

```diff
@@ -0,0 +1,164 @@
+                Some(o) => o,
+                None => LanceSinkOpts::default(),
+            },
+        })
+    }
+
+    async fn stream_into_inner(
+        &self,
+        stream: SendableRecordBatchStream,
+        mut ds: Option<Dataset>,
+    ) -> DfResult<Option<Dataset>> {
+        let mut opts = FileWriterOptions::default();
+        if self.opts.collect_all_column_stats.is_some_and(|val| val) {
+            opts.collect_stats_for_fields = stream
+                .schema()
+                .fields
+                .iter()
+                .enumerate()
+                .map(|v| v.0 as i32)
+                .collect();
+        } else if self.opts.column_stats.is_some() {
+            let colls: &Vec<String> = self.opts.column_stats.as_ref().unwrap();
+            let mut set = HashSet::with_capacity(colls.len());
+
+            for c in colls.iter() {
+                set.replace(c);
+            }
+
+            opts.collect_stats_for_fields = stream
+                .schema()
+                .fields
+                .iter()
+                .map(|f| f.name().to_owned())
+                .filter(|f| set.contains(f))
+                .enumerate()
+                .map(|v| v.0 as i32)
+                .collect();
+        }
+        let table = if self.opts.url.is_some() {
+            self.opts
+                .url
+                .clone()
+                .unwrap()
+                .join(self.loc.as_ref())
+                .map_err(|e| DataFusionError::External(Box::new(e)))?
+        } else {
+            url::Url::parse(self.loc.as_ref())
+                .map_err(|e| DataFusionError::External(Box::new(e)))?
+        };
+
+        let schema = stream.schema().clone();
+        let mut chunks = stream.chunks(32);
+        let write_opts = WriteParams {
+            mode: WriteMode::Overwrite,
+            ..Default::default()
+        };
+        while let Some(batches) = chunks.next().await {
+            let batch_iter =
+                RecordBatchIterator::new(batches.into_iter().map(|item| Ok(item?)), schema.clone());
+
+            match ds.clone() {
+                Some(mut d) => {
+                    d.append(batch_iter, Some(write_opts.clone())).await?;
+                }
+                None => {
+                    ds.replace(
+                        Dataset::write(batch_iter, table.as_str(), Some(write_opts.clone()))
+                            .await?,
+                    );
+                }
+            }
+        }
+
+        Ok(ds)
+    }
+}
+
+#[async_trait]
+impl DataSink for LanceSink {
+    async fn write_all(
+        &self,
+        data: Vec<SendableRecordBatchStream>,
+        _context: &Arc<TaskContext>,
+    ) -> DfResult<u64> {
+        let mut ds: Option<Dataset> = None;
+        for stream in data {
+            ds = self.stream_into_inner(stream, ds).await?;
+        }
+        match ds {
+            Some(ds) => Ok(ds.count_rows().await? as u64),
+            None => Ok(0),
+        }
+    }
```

Comment:
There's only one dataset it's constructed in the first call to `stream_into_inner` and used thereafter. 

_having said that_ because dataset is a handle to a lance table, this would work no matter what, because it's actively checking the metadata of the table after all writes have been complete (so in theory, (if copy to supported append and not just overwrite) we could use this to have multiple concurrent inserts.)